studies/tasks.py

```diff
@@ -39,6 +47,166 @@
+        while page := cursor.fetchmany(page_size):
+            yield from starmap(MessageTarget, page)
+
+
+def _grouped_by_user(potential_targets):
+    """Groups users with their child-study notification pairs.
+
+    Note: Depends on sorted output (hence the ORDER BY in SQL) See:
+    https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.groupby_transform
+    """
+    get_user_id = attrgetter("user_id")
+    get_child_study_pair = attrgetter("child_id", "study_id")
+    yield from (
+        (
+            user_id,
+            tuple(pair),
+        )  # Apparently, valuefunc gets wrapped to return a map object?
+        for user_id, pair in groupby_transform(
+            potential_targets, keyfunc=get_user_id, valuefunc=get_child_study_pair,
+        )
+    )
+
+
+def _deserialized(user_grouped_targets, number_of_parents: int = 100):
+    """Fill out groups with real models.
+
+    In prod, we have about 1.42 children per parent, so limiting it to 100 parents or so
+    at a time should be fine.
+    """
+    from studies.models import Study
+
+    # Cache studies entirely upfront; let the generator hold on to them.
+    study_cache = {
+        study.id: study
+        for study in Study.objects.filter(state="active", public=True).select_related(
+            "lab"
+        )
+    }
+
+    # Go in batches sized by number of parents, fetch parents and children in one SELECT
+    # query (each) before yielding deserialized groups.
+    for group_list in chunked(user_grouped_targets, n=number_of_parents):
+        user_cache = {
+            user.id: user
+            for user in User.objects.filter(id__in=[group[0] for group in group_list])
+        }
+        child_cache = {
+            child.id: child
+            for child in Child.objects.filter(
+                id__in=flatten(list(map(first, group[1])) for group in group_list)
+            )
+        }
+
+        for user_id, child_study_pairs in group_list:
+            yield (
+                user_cache.get(user_id),
+                [
+                    (child_cache.get(child_id), study_cache.get(study_id))
+                    for child_id, study_id in child_study_pairs
+                ],
+            )
+
+
+def _validated(deserialized_groups):
+    """Yield only groups with targets that satisfy criteria for their respective studies."""
+    for user, child_study_pairs in deserialized_groups:
+        valid_message_targets = []
+        for pair in child_study_pairs:
+            child, study = pair
+            eligible = get_child_eligibility_for_study(child, study)
+            if eligible:
+                valid_message_targets.append(pair)
+        if valid_message_targets:
+            yield user, valid_message_targets
+
+
+def _segmented_by_study(validated_groups):
+    for user, child_study_pairs in validated_groups:
+        yield user, dict(map_reduce(child_study_pairs, itemgetter(1), itemgetter(0)))
+
+
+@app.task
+def send_announcement_emails():
```

Comment:
Out of curiosity what is the process for actually adding the task to do this daily?

Reply:
Actually, we need to add a row in `django_celery_beat_periodictasks` - I should take care of that in a migration